primitives/runtime/src/offchain/storage_lock.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Offchain Storage Lock
+//!
+//! A persistent storage lock with a defined expire time.
+//!
+//! One use case would be collective updates of multiple data items
+//! or append / remove of i.e. sets, vectors which are stored in
+//! the offchain storage DB.
+//!
+//! ## Example:
+//!
+//! ```rust
+//! // in your off-chain worker code
+//!
+//! fn append_to_in_storage_vec<'k, T>(key: &'k [u8], T) where T: Encode {
+//!    let mut lock = StorageLock::new(b"x::lock");
+//!    if let Ok(_guard) =  lock.spin_lock() {
+//!         let acc = StorageValueRef::persistent(key);
+//!         let v: Vec<T> = acc.get::<Vec<T>>().unwrap().unwrap();
+//!         // modify `v` as desired
+//!         acc.set(v);
+//!    } else {
+//!         // the lock duration expired
+//!    }
+//! }
+//! ```
+
+use crate::offchain::storage::StorageValueRef;
+use sp_core::offchain::{Duration, Timestamp};
+use sp_io::offchain;
+
+/// Default expirey duration in milliseconds.
+const STORAGE_LOCK_DEFAULT_EXPIREY_DURATION: u64 = 30_000;
+
+/// Snooze duration before attempting to lock again in ms.
+const STORAGE_LOCK_PER_CHECK_ITERATION_SNOOZE: u64 = 100;
+
+/// A persisted guard state.
+///
+/// An in DB persistent mutex for multi access items which are modified
+/// i.e. vecs or sets.
+pub struct StorageLock<'a> {
+	// A storage value ref which defines the DB entry representing the lock.
+	value_ref: StorageValueRef<'a>,
+	// `None` implies it was already released by `fn unlock(..)`
+	locked_until: Option<Timestamp>,
+}
+
+impl<'a> StorageLock<'a> {
+	/// Create a new storage lock with [default expirey duration](Self::STORAGE_LOCK_DEFAULT_EXPIREY_DURATION).
+	pub fn new<'k>(key: &'k [u8]) -> Self
+	where
+		'k: 'a,
+	{
+		Self {
+			value_ref: StorageValueRef::<'a>::persistent(key),
+			locked_until: None,
+		}
+	}
+
+	#[inline]
+	fn try_lock_inner(&mut self, duration: Duration) -> Result<(), Option<Timestamp>> {
```

Comment:
Depending on how long it takes, this would skew the time for which the lock is actually held, hence the duration which will be used to form a deadline once the lock is acquired.

Reply:
The changes addressed this by always using a conservative deadline (= where the lock becomes stale / invalid).
For certain `Lockable` types this may be overridden given the extension trait approach.